### PR TITLE
Display resource-access events in marketplace item tester

### PR DESCRIPTION
**Pull request by kkurian**
Display resource-access events in the marketplace item tester.

Also, convert the marketplace item tester into a single-page app.

### CODE REVIEW

**Comment by kkurian** [comment]
https://trello.com/c/28O8eN9v/692-tester-capture-display-resources
https://highfidelity.fogbugz.com/f/cases/19196/marketplace-item-tester-display-resources-used

This test refers to an "asset." Example assets are provided here:
https://www.dropbox.com/sh/77y67519q7ult25/AACRkBRFkdAe6p5hRFgdTlXSa?dl=0

Test plan:

1. Open Developer > Marketplace Item Tester
2. Select "Load File"; a file picker should appear
3. Use the file picker a asset to load; any kind of asset except application is fine for this test
4. Verify: name and path of asset appears in the marketplace item tester panel
5. If the combo box displays "unknown" then select "entity" or "wearable" as appropriate; otherwise verify: the combo box automatically displays the correct kind of asset
6. Select the icon next to the trash can, which should instantiate the asset
7. Verify: the asset instantiates
8. Verify: "0 resources loaded..." changes to reporting some number of resources loaded
9. Select the maximize "+" icon next to "... resources loaded ..."
10. Verify: a text area appears and "... resources loaded ..." is replaced by a list of resources, one line each, each line containing a timestamp followed by a url (it is possible that instead of a url there is just a number -- this is okay)
11. Verify: the number of resources in the maximized view matches the number of resources reported in the minimized view
12. Select the minimize "-" icon next to the text area
13. Verify: the text area minimizes back to the "... resources loaded ..." line and that the line correctly reports the number of resources as seen in the maximized view
14. Close the marketplace item tester
15. Open the marketplace item tester (i.e. repeat Step 1)
16. The asset(s) previously loaded into the panel should still appear however all should report "0 resources loaded..." and their maximized views should show no assets loaded.
17. Repeat 6-14

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14081/HighFidelity-Beta-PR14232-89a5da8.apk

**Comment by kkurian** [comment]
[tester: capture & display resources](https://trello.com/c/28O8eN9v/692-tester-capture-display-resources)

**Comment by kkurian** [comment]
[tester: capture & display resources](https://trello.com/c/28O8eN9v/692-tester-capture-display-resources)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14081/HighFidelity-Beta-PR14232-89a5da8.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14081/HighFidelity-Beta-Interface-PR14232-89a5da8.exe

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14104/HighFidelity-Beta-PR14232-2f37a2b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14104/HighFidelity-Beta-Interface-PR14232-2f37a2b.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14127/HighFidelity-Beta-PR14232-9f91238.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14127/HighFidelity-Beta-PR14232-9f91238.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14127/HighFidelity-Beta-Interface-PR14232-9f91238.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14133/HighFidelity-Beta-PR14232-d7d49ed.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14133/HighFidelity-Beta-PR14232-d7d49ed.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14133/HighFidelity-Beta-Interface-PR14232-d7d49ed.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14169/HighFidelity-Beta-PR14232-45bb53b.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14169/HighFidelity-Beta-PR14232-45bb53b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14169/HighFidelity-Beta-Interface-PR14232-45bb53b.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14183/HighFidelity-Beta-PR14232-c6ccc21.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14183/HighFidelity-Beta-PR14232-c6ccc21.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14183/HighFidelity-Beta-Interface-PR14232-c6ccc21.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14183/HighFidelity-Beta-PR14232-c6ccc21.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14183/HighFidelity-Beta-Interface-PR14232-c6ccc21.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14186/HighFidelity-Beta-PR14232-a956677.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14186/HighFidelity-Beta-PR14232-a956677.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14186/HighFidelity-Beta-Interface-PR14232-a956677.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14186/HighFidelity-Beta-PR14232-a956677.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14186/HighFidelity-Beta-Interface-PR14232-a956677.exe

**Comment by howard-stearns** [comment]
Status: I _think_ this is now ready for testing and code review.

I'm going to try to do some thinking about whether it's worth it to pull out some of the code that is consistent with everything and consistent with where we want to go with this, but which isn't actually being used yet.

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14212/HighFidelity-Beta-PR14232-c7927c2.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14212/HighFidelity-Beta-PR14232-c7927c2.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14212/HighFidelity-Beta-Interface-PR14232-c7927c2.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14212/HighFidelity-Beta-PR14232-c7927c2.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14212/HighFidelity-Beta-Interface-PR14232-c7927c2.exe

**Comment by qlihifitest** [comment]
Currently Testing

**Comment by qlihifitest** [comment]
PASS @conklin94122

**Comment by zfox23** [comment]
Here are my high-level changes from Friday:
1. We only support 1 item at a time in the tester for now, since we can't currently tie resource requests to individual items.
2. I split the item tester QML into two items: `ItemUnderTest.qml` is the `Delegate` for the main app's `ListView`, and `MarketplaceItemTester.qml` is the main app.
3. The resource requests for each item didn't used to save across tester app sessions. Now they do. (That meant moving that text data from the QML into the JS.)
4. I fixed the build errors for MacOS and Ubuntu. At the same time, I removed all `QUrlAncestry` work (which made me sad to do, because I know we are going to have to re-add it).
I [isolated this to one commit](https://github.com/highfidelity/hifi/pull/14232/commits/c6ccc218b83f391c2bbb31982c5d504aeb7a5f0e), which might make it easier to re-add:
5. There was a bug where one of the resource observer update calls would be made twice erroneously. [This commit](https://github.com/highfidelity/hifi/pull/14232/commits/a95667792da8c8dd46d977e47349935c5c6baf08) fixes that. (Thanks to @roxanneskelly!)
6. The tester now only shows unique instances of source and URL pairs. For example, if a JSON entity requests `drum.fbx` like this:
    - get `drum.fbx` from internet
    - get `drum.fbx` from cache
   - get `drum.fbx` from cache
Only the first two lines will show up in the tester logs.
7. The "rez" button in the tester now controls starting and stopping the recording of resource requests.
8. The "rez" button for an app will now install and open (before, the app would be installed when you selected the `app.json` from disk/remote, then open with the button)

File: libraries/octree/src/Octree.cpp [hunk 1]
```diff
@@ -689,7 +689,7 @@ bool Octree::readFromFile(const char* fileName) {
 
     qCDebug(octree) << "Loading file" << qFileName << "...";
 
-    bool success = readFromStream(fileLength, fileInputStream);
+    bool success = readFromStream(fileLength, fileInputStream, "");
```
Comment (davidkelly) [inline libraries/octree/src/Octree.cpp]: "" is already the default for the 3rd parameter.  But that's fine, no need to change unless there are other changes.

File: scripts/system/marketplaces/marketplaces.js [hunk 2]
```diff
@@ -522,13 +558,19 @@ function getPositionToCreateEntity(extra) {
...
 
-function rezEntity(itemHref, itemType) {
+function defaultFor(arg, val) {
+    return typeof arg !== 'undefined' ? arg : val;
+}
+
+function rezEntity(itemHref, itemType, marketplaceItemTesterId) {
     var isWearable = itemType === "wearable";
-    var success = Clipboard.importEntities(itemHref);
+    print("!!!!! Clipboard.importEntities " + marketplaceItemTesterId);
```
Comment (davidkelly) [inline scripts/system/marketplaces/marketplaces.js]: I guess we can remove this

File: scripts/system/marketplaces/marketplaces.js [hunk 3]
```diff
@@ -877,11 +920,16 @@ var onQmlMessageReceived = function onQmlMessageReceived(message) {
     case 'checkout_rezClicked':
     case 'purchases_rezClicked':
     case 'tester_rezClicked':
-        rezEntity(message.itemHref, message.itemType);
+        print("!!!!! marketplaces tester_rezClicked");
```
Comment (davidkelly) [inline scripts/system/marketplaces/marketplaces.js]: maybe we don't need this?

File: interface/src/commerce/QmlCommerce.cpp [hunk 4]
```diff
@@ -325,8 +325,8 @@ bool QmlCommerce::installApp(const QString& itemHref) {
 
     QUrl appHref(itemHref);
 
-    auto request = DependencyManager::get<ResourceManager>()->createResourceRequest(
-        this, appHref, true, -1, "QmlCommerce::installApp");
+    auto request =
+        DependencyManager::get<ResourceManager>()->createResourceRequest(this, appHref, true, -1, "QmlCommerce::installApp");
```
Comment (kkurian) [inline interface/src/commerce/QmlCommerce.cpp]: @zfox23 Why? Your preference or following an official style guideline?

Comment (zfox23) [reply interface/src/commerce/QmlCommerce.cpp]: This is what Clang did.

**Comment by kkurian** [comment]
@davidkelly @JeffClinton I just pushed a clean-up code review that I am certain changes none of the deliverable's behavior or performance. This caused your reviews to be dismissed. LATER TODAY OR TOMORROW BUT NOT NOW I will ask you to rubber stamp this PR again. Just giving you a heads up because when this is ready for final approval it will require immediate attention. Assuming nothing else changes there will be nothing for you to review -- it will only need your re-approval. Thank you.

File: libraries/networking/src/ResourceCache.cpp [hunk 5]
```diff
@@ -338,28 +339,30 @@ QSharedPointer<Resource> ResourceCache::getResource(const QUrl& url, const QUrl&
...
-    resource->moveToThread(qApp->thread());
-    connect(resource.data(), &Resource::updateSize, this, &ResourceCache::updateTotalSize);
-    {
-        QWriteLocker locker(&_resourcesLock);
-        _resources.insert(url, resource);
+    if (!resource) {
+        resource = createResource(
+            url,
+            fallback.isValid() ?  getResource(fallback, QUrl()) : QSharedPointer<Resource>(),
+            extra);        resource->setSelf(resource);
```
Comment (roxanneskelly) [inline libraries/networking/src/ResourceCache.cpp]: nit: missing carriage return

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14241/HighFidelity-Beta-PR14232-8592a8f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14241/HighFidelity-Beta-Interface-PR14232-8592a8f.dmg

**Comment by howard-stearns** [comment]
As I understand it, this is being held open as DNM only to facilitate review, particularly outside review.

**Comment by kkurian** [comment]
@davidkelly I just resolved merge conflicts... so care to approve again? Third time is a charm?

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/14260/HighFidelity-Beta-PR14232-a2a6acd.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14260/HighFidelity-Beta-PR14232-a2a6acd.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14260/HighFidelity-Beta-Interface-PR14232-a2a6acd.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14260/HighFidelity-Beta-PR14232-a2a6acd.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14260/HighFidelity-Beta-Interface-PR14232-a2a6acd.exe

File: libraries/shared/src/ResourceRequestObserver.cpp [hunk 6]
```diff
@@ -0,0 +1,28 @@
+//
+//  ResourceAccessMonitor.h
```
[the remark this answers is not in the document]
Comment (kkurian) [reply libraries/shared/src/ResourceRequestObserver.cpp]: Will do in the next rev. Don't want to derail existing approvals by making this change.

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14315/HighFidelity-Beta-PR14232-e7d1c28.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/14315/HighFidelity-Beta-Interface-PR14232-e7d1c28.dmg

**Comment by kkurian** [comment]
Marked all of Huffman's comments as resolved. All have been noted and will be addressed later, in https://trello.com/c/wUsoZ5Da/693-tester-capture-referrers

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14315/HighFidelity-Beta-PR14232-e7d1c28.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/14315/HighFidelity-Beta-Interface-PR14232-e7d1c28.exe

**Comment by qlilaseguinolaza** [comment]
Currently testing

**Comment by qlilaseguinolaza** [comment]
PASS @conklin94122

**Pull request merged**